Proposal: Static types

bentsherman opened the pull request:
This PR is a showcase of the proposed syntax for static types in Nextflow.

While I started with the goal of simply adding type annotations and type checking, I realized that many aspects of the language needed to be re-thought in order to provide a consistent developer experience. Some of these things can be done now, but I suspect they will be more difficult without static types, so I have tried to show them in their "best form" in this PR.

### Changes

- **Update to Nextflow 25.04.** See #347 for details.

- **Type annotations.** The following declarations can be annotated with a type:
  - workflow params/outputs
  - workflow takes/emits
  - process inputs/outputs
  - function parameters/return
  - local variables (generally not needed)

  Nextflow will use these type annotations to infer the type of every value in the workflow and make sure they are valid.

  The main built-in types are:
  - `Integer`, `Float`, `Boolean`, `String`: primitive types
  - `Path`: file or directory
  - `List<E>`, `Set<E>`, `Bag<E>`: collections with various constraints on ordering and uniqueness
  - `Map<K,V>`: map of key-value pairs
  - `Channel<E>`: channel (i.e. queue channel)
  - `Value<V>`: dataflow value (i.e. value channel)

- **Records, optional types.** Types can be composed in several ways to facilitate domain modeling:
  - records: a combination of named values, e.g. a sample is a meta map AND some files:
    `record Sample { meta: Map ; files: List<Path> }`
  - optionals: any type can be suffixed with `?` to denote that it can be null (e.g. `String?`), otherwise it should never be null

- **Define pipeline params in the main script.** Each param has a type. Complex types can be composed from collections, records, and enums. Rather than specifying a particular input format for input files, simply specify a type and Nextflow will use the type like a schema to transparently load from any source (CSV/JSON/etc). Config params are defined separately in the main config. `nextflow_schema.json` remains unchanged but will be partially generated from the main script / config.

- **Only use params in entry workflow.** Params are not known outside the entry workflow. Pass params into processes and workflows as explicit inputs instead.

- **Processes are just functions.** Instead of calling a process directly with channels, use operators and supply the process name in place of an operator closure:

  ```groovy
  // execute FASTQC in parallel on each input file
  channel.fromPath( "inputs/*.fastq" ).map(FASTQC)

  // execute ACCUMULATE sequentially on each input file
  // (replaces experimental recursion)
  channel.fromPath( "inputs/*.txt" ).reduce(ACCUMULATE)
  ```

- **Simple operators.** Use a simple and composable set of operators:
  - `collect`: collect channel elements into a collection (i.e. bag)
  - `cross`: cross product of two channels
  - `filter`: filter a channel based on a condition
  - `flatMap`: nested scatter
  - `groupTuple`: nested gather
  - `join`: relational join of two channels (i.e. horizontal)
  - `map`: transform a channel
  - `mix`: concatenate multiple channels (i.e. vertical)
  - `reduce`: accumulate each channel element into a single value
  - `scan`: like `reduce` but emit each intermediate value
  - `subscribe`: invoke a function for each channel element
  - `view`: print each channel element

### Benefits

- **Well-defined workflow inputs.** Workflow inputs are explicitly defined alongside the entry workflow as a set of name-type pairs (i.e. a record type). Complex params can be loaded transparently from any source (file, database, API, etc) as long as the runtime supports it. The JSON schema of a param is inferred from the param's type.

- **Well-defined workflow outputs.** Workflow outputs are explicitly defined as a set of name-type pairs (i.e. a record type). Each output can create an index file, which is essentially a serialization of a channel to external storage (file, database, API, etc), and each output can define how its published files are organized in a directory tree. The JSON schema of an output is inferred from the output's type.

- **Make pipeline import-able.** Separating the "core" workflow (i.e. `SRA`) from params and publishing makes it easy to import the pipeline into larger pipelines. See https://github.com/bentsherman/fetchngs2rnaseq for a more complete example.

- **Simpler dataflow logic.** Processes are called like an operator closure, generally with a single channel of maps. where the map keys correspond to the process inputs. Additional inputs can be provided as named args. As a result, the amount of boilerplate in the workflow logic and process definition is significantly reduced.

- **Simpler operator library.** With a minimal set of operators, users can easily determine which operator to use based on their needs. The operators listed above are statically typed and pertain only to stream operations.

- **Simpler process inputs/outputs.** Process inputs/outputs are declared in the same way as workflow takes/emits and pipeline params/outputs, instead of the old custom type qualifiers. Inputs of type `Path` are automatically staged. Thanks to the simplified dataflow logic described above, tuples are generally not needed.

### Extra Notes

This proposed syntax will be enabled by the following internal improvements:

- New script/config parser, which enables us to evolve the Nextflow language into whatever we want, without being constrained by Groovy syntax (though it still must compile to Groovy AST).

- Static analysis, which can infer the type of every value based on the declared types of pipeline/workflow/process inputs.

- Automatic generation of JSON schemas for workflow params and outputs based on type annotations. Preserves support for external tools like Seqera Platform, and lays the groundwork to transparently support different connectors (CSV/JSON file, HTTP API, SQL database, etc).

bentsherman commented on workflows/sra/main.nf:
```diff
@@ -29,164 +37,171 @@ include { FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS } from '../../subworkflow
+            |> filter { meta ->
+                getDownloadMethod(meta, download_method) == 'aspera'
+            }                                                   // Channel<Map>
+            |> map { meta ->
+                def sample = new Sample( meta, meta.fastq_aspera.tokenize(';').take(2).collect( name -> file(name) ) )
+                ASPERA_CLI ( sample, 'era-fasp', aspera_cli_args )
+            }                                                   // fastq: Channel<Sample>, md5: Channel<Sample>
```
@mahesh-panchal to your point about dynamic args, I think we can do even better in DSL3:

```groovy
|> map { meta ->
  def sample = new Sample( /* ... */ )
  ASPERA_CLI ( sample, 'era-fasp', "${meta.key}" )
}
```

Because we call the process in a `map` operator explicitly (currently it is implied), we can control how the process is invoked for each task within the operator closure, instead of passing multiple queue channels.

mahesh-panchal replied:
Yes, this specifically is much better. Treating processes like functions is soooo much better since there's no implicit transformation stuff going on with all the singleton/queue channel stuff. It has to be formed and then `map`ped. And `tuple`s disappear too, except in channels (?).

Actually I think what's worrying me about this syntax is the mixing of input types. An input could be a channel (e.g. `MULTIQC_MAPPINGS_CONFIG ( mappings )` lower down ) or it could be an input set (e.g. this dynamically defined `Sample`). This is already confusing to new comers where we commonly see people trying to use channels inside `map`, `branch`, etc.
I guess one could explain the second option as passing dynamically defined singleton channels.

bentsherman replied:
In this proposal, there are no "value" channels, only queue channels and regular values. So the `MULTIQC_MAPPINGS_CONFIG ( mappings )` is no different because `mappings` is just a value. It may be an async value, and Nextflow might represent it as a value channel under the hood, but to the user it should be indistinguishable from a regular value

In other words, you can not call a process with a channel, only values

bentsherman commented on modules/local/sra_to_samplesheet/main.nf:
```diff
@@ -1,25 +1,39 @@
-    val pipeline
-    val strandedness
-    val mapping_fields
+    List<Map> sra_metadata
```
yes exactly

maxulysse replied:
`List<Map>` syntax is a bit hard/complex

mahesh-panchal replied:
It's groovy syntax for generics: https://groovy-lang.org/objectorientation.html#generics

bentsherman replied:
also it's a list of maps so you can't get much simpler than `List<Map>`

bentsherman commented on main.nf:
```diff
@@ -33,14 +33,29 @@ include { PIPELINE_COMPLETION     } from './subworkflows/local/utils_nfcore_fetc
     // WORKFLOW: Download FastQ files for SRA / ENA / GEO / DDBJ ids
     //
-    SRA ( ids )
+    SRA (
+        ids,
+        params.ena_metadata_fields ?: '',
+        params.sample_mapping_fields,
+        params.nf_core_pipeline ?: '',
+        params.nf_core_rnaseq_strandedness ?: 'auto',
+        params.download_method,
+        params.skip_fastq_download,
+        params.dbgap_key,
+        params.aspera_cli_args,
+        params.sra_fastq_ftp_args,
+        params.sratools_fasterqdump_args,
+        params.sratools_pigz_args,
+        params.outdir
+    )
```
You could make a record type 😄

mahesh-panchal replied:
An option I'd considered was to have those params as part of the record that supplied meta and files to stage.

bentsherman replied:
Added `SraParams` record type in latest updates

bentsherman commented on modules/local/aspera_cli/main.nf:
```diff
@@ -8,16 +10,21 @@ process ASPERA_CLI {
-    tuple val(meta), path("*md5")     , emit: md5
-    path "versions.yml"               , emit: versions
+    Sample fastq    = new Sample(meta, path("*fastq.gz"))
+    Sample md5      = new Sample(meta, path("*md5"))
```
I've considered it. The proposed syntax is the most basic form that matches how variables are declared in general.

I guess you don't really need the output type on the left if it can always be inferred from the right-hand side.

Alternatively, since we always call a process with a single input -> single output with this proposed syntax, instead of channels, we could also just specify each record element as a separate input/output and bundle them into records in the workflow as needed. But that might be unwieldy for records with many elements.

bentsherman replied:
Did some refactoring today. Since process calls are so much more flexible now, I think we can simplify a few things here:
- the output type can be omitted, inferred from the right-hand side
- if there is only one output, the output name can be omitted because the process will just return that value directly
- if there are multiple outputs, the process will return an "implicit" record type, similar to the process `.out` but with single values instead of channels.
- I removed the use of `Sample` from all processes since it doesn't add much value. Instead I only bundle some things into `Sample`s at the workflow level where it makes sense. Of course for larger record types it might be different.

Hopefully that makes the types less daunting as well. Basically the only place where they need to be explicitly declared is for function/process/workflow inputs.

bentsherman commented on workflows/sra/main.nf:
```diff
@@ -29,164 +37,171 @@ include { FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS } from '../../subworkflow
-            }
-            .set { ch_sra_reads }
+    ids                                                         // Channel<String>
+        //
+        // MODULE: Get SRA run information for public database ids
+        //
+        |> map { id ->
+            SRA_IDS_TO_RUNINFO ( id, ena_metadata_fields )
+        }                                                       // Channel<Path>
+        //
+        // MODULE: Parse SRA run information, create file containing FTP links and read into workflow as [ meta, [reads] ]
+        //
+        |> map(SRA_RUNINFO_TO_FTP)                              // Channel<Path>
+        |> set { runinfo_ftp }                                  // Channel<Path>
+        |> flatMap { tsv ->
+            splitCsv(tsv, header:true, sep:'\t')
+        }                                                       // Channel<Map>
+        |> map { meta ->
+            meta + [single_end: meta.single_end.toBoolean()]
+        }                                                       // Channel<Map>
+        |> unique                                               // Channel<Map>
+        |> set { sra_metadata }                                 // Channel<Map>
+
```
Indeed. A lot of moving pieces have to come together to make this work. Does it make sense to you how I am calling the process like a function in the map operator?

mahesh-panchal replied:
It took a second to process, but this is soooo much better. This deals with what I wanted much better than how I initially thought about from the current syntax.

maxulysse replied:
what is `|>` used for?

maxulysse replied:
ooh, I see, sorry

bentsherman replied:
Reverted the pipe operator in favor of dots, but the overall structure is essentially the same. The key improvement is moving the process call into the map operator, which allows you to have longer operator chains without having to break it up just to call a process

mahesh-panchal commented on main.nf:
```diff
@@ -33,14 +33,29 @@ include { PIPELINE_COMPLETION     } from './subworkflows/local/utils_nfcore_fetc
+        params.ena_metadata_fields ?: '',
+        params.sample_mapping_fields,
+        params.nf_core_pipeline ?: '',
+        params.nf_core_rnaseq_strandedness ?: 'auto',
```
This was one of the issues of the original method Harshil implemented. There were various places where parameter settings could sneak in. It's easy enough to make it a guideline though that defaults must be in the config, especially now with schema validation.

bentsherman replied:
yes it would probably be better to just put these defaults in the schema, perhaps allow the workflow `take:` definition to define default values if it helps modularity

mahesh-panchal commented on main.nf:
```diff
@@ -86,6 +102,11 @@ workflow {
 }
 
+publish {
+    directory params.outdir
```
Can we have the syntax have an `=` or `:` here for readability? Or is this a function?
```
directory(params.outdir)
```

bentsherman replied:
it is a function call under the hood, so you could use parentheses here. it is the same syntax as process directives.

personally I would rather put these settings in the config file because they seem more like config, but Paolo prefers this form for now

mahesh-panchal commented on modules/nf-core/sratools/fasterqdump/main.nf:
```diff
@@ -8,21 +10,26 @@ process SRATOOLS_FASTERQDUMP {
 
-    when:
-    task.ext.when == null || task.ext.when
+    topic:
+    [ task.process, 'sratools', eval("fasterq-dump --version 2>&1 | grep -Eo '[0-9.]+'") ] >> 'versions'
+    [ task.process, 'pigz', eval("pigz --version 2>&1 | sed 's/pigz //g'") ] >> 'versions'
```
I kind of like this, but I dislike the name `topic`. I don't feel like the word communicates what it's function is. 

It would also be nice if we could supply a regex to validate what should be returned by the `eval` for some fast fail behavior when there's extra stuff being emitted. Where would one define a global variable pattern? E.g. 
```
def SOFTWARE_VERSION = /\d+.../
def SHASUM = /\w{16}/ 
```
Or maybe this should be a class? like you can `filter { Number }`.

bentsherman replied:
`topic` is a term from stream processing, used to collect related events from many different sources. in this case we are sending the tool version info to a custom "versions" topic, then the workflow reads from that topic to build the versions yaml file.

`eval` is just a function defined in the output / topic scope, so you could wrap it in a custom validation function:
```groovy
def validate( pattern, text ) {
  // ...
}

// ...
  topic:
  validate( /foo/, eval('...') ) >> 'versions'
```

mahesh-panchal commented on assets/schema_input.yml:
```diff
@@ -0,0 +1,12 @@
+$schema: http://json-schema.org/draft-07/schema
```
Changing to YAML is good idea I think, particularly for debugging.

mahesh-panchal commented on modules/local/sra_fastq_ftp/main.nf:
```diff
@@ -10,15 +11,20 @@ process SRA_FASTQ_FTP {
-    tuple val(meta), path("*fastq.gz"), emit: fastq
-    tuple val(meta), path("*md5")     , emit: md5
-    path "versions.yml"               , emit: versions
+    Sample fastq    = new Sample(meta, path("*fastq.gz"))
+    Sample md5      = new Sample(meta, path("*md5"))
```
This is something we're really going to need to flesh out when it comes to (nf-core) outputs. 
The `Sample` record here is just being used to replace `tuple`, but with the ability to create records, perhaps outputs will be much cleaner. It's still not clear to me how this will be used in a more complex workflow that's using `combine/join` on subsets of record fields.

bentsherman replied:
This record is simple enough that you could keep it as a tuple and the type would be `Tuple2<Map,List<Path>>`, of course that gets messy as you add more members

Operators like `combine`, `join`, `groupTuple` are designed to work with tuples, so they'll need to be extended to support records and maybe also maps. There is an ongoing discussion for that here: https://github.com/nextflow-io/nextflow/issues/3108 . The `by` option needs to be a string or even a closure so that you can select an arbitrary path within a data structure

bentsherman replied:
This is actually related to that plugin you were writing. We should probably review that and see if it might fit into this proposal re: extending those operators

mahesh-panchal replied:
I stopped implementing it after writing the part for `groupTuple`. I swapped to using functions because it wasn't worth the effort.
The functions are here: https://github.com/NBISweden/Earth-Biogenome-Project-pilot/blob/dev/modules/local/functions.nf

bentsherman replied:
Right, I think those operators can be extended to support that use case by allowing `by` to be a closure instead of a tuple index. The result will also need to be more generic instead of being optimized for tuples, e.g. `join` should simply return a 3-tuple of key, left value, right value, then you could use `map` to shape it however you want.

bentsherman replied:
Records and tuples have been removed from the process definitions as they are generally not needed anymore in the proposed syntax. I have confirmed that the proposed set of minimal operators can be designed to work with any type, not just tuples.

mahesh-panchal commented on workflows/sra/main.nf:
```diff
@@ -29,164 +37,171 @@ include { FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS } from '../../subworkflow
-    }
+    sample_mappings = sample_mapping_fields
+        ? MULTIQC_MAPPINGS_CONFIG ( mappings )              // Path
+        : null
```
So this replaces `Channel.empty()`?

bentsherman replied:
yes

bentsherman replied:
to be clear, `Channel.empty()` can still be used as a fallback value for a channel, but in this case, `sample_mappings` is just a value, so we just use null here

mahesh-panchal commented on workflows/sra/main.nf:
```diff
@@ -29,164 +37,171 @@ include { FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS } from '../../subworkflow
+        |> Channel.topic                                    // Channel<Tuple3<String,String,String>>
+        |> softwareVersionsToYAML                           // Channel<String>
+        |> collect(sort: true)                              // List<String>
+        |> exec('SOFTWARE_VERIONS') { versions ->
```
Wait, so we're getting inline definition of processes? I need to think on this.

bentsherman replied:
it's just an idea. I have implemented a basic version in [nf-boost](https://github.com/bentsherman/nf-boost) if you'd like to play with it

there's this whole issue with operators that create files like `splitCsv` and `collectFile`, because those files don't get tracked the same way as process outputs. so for example the automatic cleanup in nf-boost can't delete these files because it can't track them. part of this proposal which I didn't mention is to remove all operators that create files and provide functions in their stead. `splitCsv` can already be used as a regular function, and in nf-boost I have proposed `mergeText` as a replacement for `collectFile`.

so instead of e.g. the `collectFile` operator, you would call `mergeText` in an `exec` process. but having to write the process definition might feel annoying / verbose compared to an operator. so I had this idea of an `exec` operator which allows you to define an inline exec process, kinda like a closure vs function. this would allow the files to be tracked and cleaned up, without being much more verbose than an operator.

it works quite well in nf-boost. you'll even see a `SOFTWARE_VERSIONS` process in the console output. I'm just not 100% sure if it's really needed

mahesh-panchal replied:
Yes, I've only recently begun to do this now since I've explored `exec:` a bit more. I ran into the file staging issue and such, but I think it's a good idea to move these file operations to processes. The main issue I think is that we're not introduced to this concept early enough in our learning. If using native processes is part of our training early on like using scripts are then there will be no problem adopting this, and shouldn't be any need for an `exec` operator. As many from my sector come from background of writing bash/python scripts that's usually what they'll reach for, whereas if they're presented "this is how to do it", they'll likely use it as being best practice.

bentsherman replied:
All uses of collectFile were replaced with index files via workflow outputs, so the inline exec process isn't needed anymore. For now I'm leaning towards not adding it, unless someone comes up with some killer use case

mahesh-panchal commented on workflows/sra/main.nf:
```diff
@@ -29,164 +37,171 @@ include { FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS } from '../../subworkflow
+        |> softwareVersionsToYAML                           // Channel<String>
+        |> collect(sort: true)                              // List<String>
+        |> exec('SOFTWARE_VERIONS') { versions ->
+            def path = task.workDir.resolve('nf_core_fetchngs_software_mqc_versions.yml')
```
It's not explained in the docs I think, but what exactly is the benefit of using `resolve` over `file("$task.workDir/filename.ext")`

bentsherman replied:
the dynamic string will cast `task.workDir` from a path to a string, and if the path is not a local file, the URL scheme will be lost, e.g. `s3://work/dir` becomes `/work/dir`. might not matter in this case because `task.workDir` will always be local for an `exec` process.

also, the dynamic string could result in repeated slashes e.g. `/work/dir//foo.txt` which apparently can cause validation errors with certain systems https://github.com/nextflow-io/nextflow/issues/4486

mahesh-panchal commented on workflows/sra/nextflow.config:
```diff
@@ -1,8 +0,0 @@
-includeConfig "../../modules/local/multiqc_mappings_config/nextflow.config"
```
Yep. Do not want.

mahesh-panchal commented on subworkflows/local/utils_nfcore_fetchngs_pipeline/main.nf:
```diff
@@ -75,23 +75,22 @@ workflow PIPELINE_INITIALISATION {
-    } else {
+    input = file(input)
+    if (!isSraId(input))
         error('Ids provided via --input not recognised please make sure they are either SRA / ENA / GEO / DDBJ ids!')
```
Can we have a set of functions for reporting errors, tips, warnings, etc to the user without reporting script line number? As in, there should be a distinction between error messages generated for the user, and error messages generated for the developer.

And ideally something that doesn't put something into a channel if the channel is empty.

bentsherman replied:
think there is an ongoing discussion for that here: https://github.com/nextflow-io/nextflow/issues/4937

in any case, it can be done independently of these language improvements

mahesh-panchal commented on subworkflows/nf-core/fastq_download_prefetch_fasterqdump_sratools/main.nf:
```diff
@@ -7,33 +7,39 @@ include { SRATOOLS_FASTERQDUMP        } from '../../../modules/nf-core/sratools/
+            //
+            // Prefetch sequencing reads in SRA format.
+            //
+            input = SRATOOLS_PREFETCH ( input, ncbi_settings, dbgap_key )
```
I don't like this. 
```
|> operator { var -> 
    var = process1( var, ... )
    process2( var, ... )
}
```
This mixing is confusing. 
It should be:
```
|> operator { var -> process1(var, ...) }
|> operator { var -> process2(var, ...) }
```

bentsherman replied:
you can keep them separate if you want. I combined them here mainly to show that it's possible. they are just functions after all, so why not be able to compose them?

mahesh-panchal commented on workflows/sra/main.nf:
```diff
@@ -29,164 +37,171 @@ include { FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS } from '../../subworkflow
-        .set { ch_mappings }
+    sra_metadata                                            // Channel<Map>
+        |> collect                                          // List<Map>
+        |> { sra_metadata ->
```
So we don't need `map` if we can just supply closures? Does `map` have a purpose then?

bentsherman replied:
the pipe into closure is a shorthand for this:
```groovy
    index_files = SRA_TO_SAMPLESHEET (
        sra_metadata |> collect, // a.k.a. collect(sra_metadata)
        nf_core_pipeline,
        nf_core_rnaseq_strandedness,
        sample_mapping_fields
    )
```

it's a convenient way to keep the pipeline going when you can't express the step as a curried function call. in this case, I want to supply some extra arguments to `SRA_TO_SAMPLESHEET`, so I can use the closure to customize that function call instead of breaking up the pipeline.

actually I'd like to be able to curry the process call just like an operator:
```groovy
    sra_metadata
        |> collect
        |> SRA_TO_SAMPLESHEET (
                nf_core_pipeline,
                nf_core_rnaseq_strandedness,
                sample_mapping_fields
        )
        |> set { index_files }
```

in any case, it's important to understand that the result of `sra_metadata |> collect` is a value (a list of meta maps), not a value channel. you can't use operators like `map` on a value here, only on channels. there are no more value channels, only queue channels

mahesh-panchal replied:
OK. So my general confusion is around the point it's either a stream or value, vs it always being a stream. 

Does this mean operators like `transpose` will be redefined, since it'll be problematic to distinguish between a stream vs a value, and what follows `|>` could be either a Channel operator or Collection function.

bentsherman replied:
in order for `|>` to work with anything without causing ambiguities, everything needs to be typed. I've gone back and forth on whether to allow operators to accept list inputs and "cast" them to channels, but ultimately I think I would prefer to force the user to be explicit. it's also not that hard:
```groovy
1..10 
  |> Channel.of // it's just an extra line
  |> map { /* ... */ }
```

I think this makes it perfectly clear what can go into an operator: only channels (i.e. queue channels). Anything else can be converted into a channel beforehand using `Channel.of()` or `Channel.fromList()`. So no, we won't need to change operators like `transpose`, and any operator that currently returns a value channel like `collect` will just return a regular value.

You can also clearly distinguish between the `List::transpose()` method and the `transpose` operator:
```groovy
[ 1, 2, 3 ].transpose()
[ 1, 2, 3 ] |> Channel.of |> transpose
```

Note that operators can no longer be called using the dot syntax, and you can't use `|>` to call an object method, only standalone functions.

mahesh-panchal replied:
> Note that operators can no longer be called using the dot syntax, and you can't use |> to call an object method, only standalone functions.

This helps with transparency and readability a lot.

bentsherman replied:
Also, I commented the expected type of each line off to the right so that you can tell whether something is a channel or value. Ideally the IDE tooling will be able to show these type hints in the editor

mahesh-panchal replied:
My issue with understanding the comments was that my mental model was still incorrect at the time of reading it, so it just caused confusion. It makes much more sense now my mental model is corrected.

bentsherman replied:
I ended up not needing this pipe-into-closure syntax since I can just curry the function call. Likely won't add it for now unless I encounter more use cases for it.

I concluded it's better to explicitly convert collections to channels rather than doing an implicit cast. Better for the developer to communicate their intent

mahesh-panchal commented on workflows/sra/main.nf:
```diff
@@ -29,164 +37,171 @@ include { FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS } from '../../subworkflow
+            |> filter { meta ->
+                getDownloadMethod(meta, download_method) == 'sratools'
+            }                                                   // Channel<Map>
+            |> map { meta ->
+                new Tuple2<Map,String>( meta, meta.run_accession )
+            }                                                   // Channel<Tuple2<Map,String>>
+            |> FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS (
+                dbgap_key ? file(dbgap_key, checkIfExists: true) : [],
+                sratools_fasterqdump_args,
+                sratools_pigz_args )                            // Channel<Sample>
```
So how does this compose inputs? Anything that's piped in is taken as the first channel, otherwise we need to use `map`?

This would otherwise be:
```
|> map { meta ->
                def tuple2 = new Tuple2<Map,String>( meta, meta.run_accession )
                FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS (
                    tuple2,
                    dbgap_key ? file(dbgap_key, checkIfExists: true) : [],
                    sratools_fasterqdump_args,
                    sratools_pigz_args )
    }
```

I'm not sure I'm liking the flexibility in the new syntax. This makes readability harder in my opinion.

bentsherman replied:
yes, the pipe source becomes the first argument in the call. this is already how it works for operators, so just extending it to functions / processes / workflows in general

but you can't call a workflow in an operator because a workflow itself contains dataflow logic. a process is more like a regular function, which is why it can be called anywhere within a workflow.

I suspect that this syntax is easier to understand for someone new to Nextflow, but possibly harder to someone used to DSL2. People have learned a lot of things in order to cope with the complexity of dataflow logic, which will need to be unlearned

bentsherman replied:
this example is actually the inverse of the other one, so it can also be written as:
```groovy
    sra_metadata
        // ...
        |> { sra_metadata ->
            FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS (
                sra_metadata,
                dbgap_key ? file(dbgap_key, checkIfExists: true) : [],
                sratools_fasterqdump_args,
                sratools_pigz_args )
        }
````

mahesh-panchal replied:
> but you can't call a workflow in an operator because a workflow itself contains dataflow logic. a process is more like a regular function, which is why it can be called anywhere within a workflow.

Is there a reason this has to be the case? At the moment workflows seem just to be channel zip-ties, and I think many would really like them to be more like functions

bentsherman replied:
That's how I used to feel as well, but it doesn't really make sense in general. Calling a workflow in a `map` operator implies that you want the workflow to independently process each value from the input channel, like a process. But what if the workflow has dataflow logic like `groupTuple` and `reduce`? Then it needs to operate on the channel as a whole, not just the individual values.

Now there is a special case, which is a workflow that only calls processes and certain operators like `map` and `filter`, for example:
```groovy
workflow FOO {
  take: input
  main: input |> map(PROC1) |> map(PROC2) |> set { out }
  emit: out
}
```

This workflow could in theory be called within an operator because it never needs the entire channel, each value is processed independently. But in that case, it could just be an operator closure!
```groovy
workflow {
  // closure equivalent to workflow FOO
  input |> map { val ->
    PROC2(PROC1(val))

    // can use pipes here btw
    // val |> PROC1 |> PROC2
  }
}
```

mahesh-panchal replied:
> But what if the workflow has dataflow logic like `groupTuple` and `reduce` ? Then it needs to operate on the channel as a whole, not just the individual values.

I'm not convinced of this. If I knew that the subworkflow acted more like a function, then I would expect these operators only to work on the subset I passed as input and not everything. I have to admit though I don't have much experience with scatter/gather implementations so outside of the naive implementation I haven't thought about it a lot.

bentsherman replied:
Thinking about this more, I guess you could treat workflows like functions and execute each workflow on an independent set of inputs. For example you could have a channel of channels, map it with a workflow, then each workflow invocation operates on one channel. This is actually something that has been requested before.

But I suspect it would add a lot of complexity without much benefit over what is already possible. I'll have to think on it though, maybe it'll become clearer after the first round of development

bentsherman replied:
After more consideration and consulting some of my engineering colleagues, I am increasingly convinced that Nextflow's "flat" dataflow model is essential to keeping the dataflow syntax simple. We could have nested parallelism (e.g. channels of channels, operators calling workflows), but we wouldn't be able to implement some of the simplifications showcased here. Nextflow would be much more like a general-purpose language with all the same complexities around concurrency.

This is also why we use `flatMap` and `groupTuple` to perform scatters and gathers, rather than a more explicit scatter/gather like in WDL. These operators keep the dataflow flat, even with multiple nested scatters. The downside is that scattered elements lose their parent context, so you have to retain that context explicitly in the form of a "group key" if you want the gather to be efficient:

```groovy
ch_samples
  | flatMap { meta, files ->
    files.collect { file -> [ groupKey(meta, files.size()), file ] }
  }
  | PROCESS
  | groupTuple(by: 0)
  | map { meta, files -> [ meta.groupTarget, files ] }
```

I can simplify this syntax somewhat with the group key handling, but at the end of the day, the group size(s) must be preserved some way or another, otherwise the gather step cannot emit anything until all values are received. This is a consequence of flat dataflow.

Considering the typical users and use cases for Nextflow, I think this is a sensible trade-off. We'll see how I feel after the first iteration.

bentsherman replied:
Simplified sketch with static types:

```groovy
ch_samples
  .flatMap { (meta, files) ->
    meta.size = files.size()
    files.collect { file -> (meta, file) }
  }
  .map(PROCESS)
  .groupBy { (meta, _) -> (meta, meta.size) }
```

mahesh-panchal reviewed (commented):
I really like the pipe replacement `|>` is easier to see, and provides some visual directionality, helping readability. 

Something extra: How about also shifting:
```
workflow {
    workflow.onComplete {
    }
}
```
to
```
workflow {
    onStart:
    ...

    take:
    ...

    main:
    ...
    
    onComplete:
    ...
}
```

There are some things I really like here, but I have reservations about other stuff like how channels are obfuscated with their channel values, and process outputs

samuell commented:
FWIW, I have a tiny feedback, that has came up after the previous discussion (which I wasn't aware of):

[The fair keyword](https://www.nextflow.io/docs/latest/process.html#fair), describes what is to my knowledge very often called "FIFO" (First-In, First-Out) in other contexts, and might have been a clearer name? (That said, perhaps not worth the change...)

bentsherman commented:
@samuell I would say, just submit an issue for that, it is more of an API change than a syntax change

samuell commented:
Reading through the suggestion in more detail now, I'm a little concerned about this one:

> The DAG will be constructed at compile-time instead of run-time, which will allow the DAG to be more comprehensive -- include params and how they connect to processes, include conditional pipeline code (e.g. if-else statements), allow nextflow inspect to list every container that might possibly be used, etc

In my experience, there are some use cases that require run-time generated DAGs, for example when initiating pipeline structure based on values extracted as part of the workflow.

This is common e.g. in machine learning, where you might run hyper-parameter tuning, which generates values which are send to initialize downstream processes, but that might potentially also influence how the DAG is generated downstream.

I've been writing about it before: https://bionics.it/posts/dynamic-workflow-scheduling

Not sure how well this applies here, but want to raise the flag about it, since it is a real limitation we have been running into with other pipeline systems (Luigi).

EDIT: Actually, I guess since we are almost definitely talking about the DAG of processes and not the DAG of tasks, a compile-time DAG would still not rule out all of dynamic scheduling (Since the dataflow paradigm of Nextflow does dynamic task scheduling inherently). Still, it seems some cases of dynamic scheduling might be affected; those that require the process DAG structure to be defined based on outcomes of previous computations.

maxulysse commented on modules/local/aspera_cli/main.nf:
```diff
@@ -8,16 +10,21 @@ process ASPERA_CLI {
+    Sample md5      = new Sample(meta, path("*md5"))
+
+    topic:
+    [ task.process, 'aspera_cli', eval('ascli --version') ] >> 'versions'
```
why brackets?

mahesh-panchal replied:
My guess is that this could be any object but a `List` is easy to process

bentsherman replied:
yes it's just a list literal, could be any expression

mahesh-panchal replied:
Are there reasons for the above over 
```groovy
    'versions' << [ task.process, 'aspera_cli', eval('ascli --version') ]
```
This is more consistent with add/append isn't it?

bentsherman replied:
that could also work

bentsherman commented:
@samuell yes I'm talking about the process i.e. "abstract" DAG, which Nextflow already constructs before executing the pipeline. But it has to execute the script in order to do this which limits its usefulness.

mahesh-panchal commented:
How come there are new keywords (`let`, `fn`, etc)? What's the difference to `def`?

bentsherman commented:
Just another idea to consider. With a formal grammar, we don't have to adhere so closely to Groovy, we can make whatever syntax we want, as long as it can be translated to Groovy AST. So as a demonstration I have replaced `def` with more specific keywords: `fn` for function defintion, `let` for variable that can't be reassigned, `var` for variable that can be reassigned (essentially `def` vs `final` in Groovy).

Notice I also changed how types are specified: `<name>: <tyoe>` instead of `<type> <name>`, which I like personally because it emphasizes the semantic name over the type which is optional.

mahesh-panchal commented:
Is it going to be problematic if people combine groovy and this grammar? For example in `exec:` blocks.

bentsherman commented:
It would apply to all Nextflow code, including `exec:` blocks

mahesh-panchal commented:
I understood it would apply to all, but my question was really if it was possible that people could mix grammars and if so what would happen: e.g.
```
exec:
let some_var = do_stuff
def another_thing = do_other_stuff
```

bentsherman commented:
We would either drop `def` in the next DSL version (a hard cut-off) or support it temporarily with a compiler warning

bentsherman commented on modules/nf-core/sratools/fasterqdump/main.nf:
```diff
@@ -8,48 +8,43 @@ process SRATOOLS_FASTERQDUMP {
     script:
-    def args = task.ext.args ?: ''
-    def args2 = task.ext.args2 ?: ''
+    def args_fasterqdump = task.ext.args_fasterqdump ?: ''
+    def args_pigz = task.ext.args_pigz ?: ''
```
Rethinking the approach to ext args. People seem to really like being able to set these args from the config without the hassle of custom params and passing them through the workflow logic. If nothing else, I think we can keep supporting the ext approach, even with static types, so that it doesn't become a roadblock.

One way we could evolve the ext approach is to allow a process to declare additional params/args in a separate section:
```groovy
  input:
  // ...

  params:
  args_fasterqdump: String
  args_pigz: String
  prefix: String
```

These are essentially process-level params, i.e. inputs that are passed in through the config rather than the pipeline code.

The config would look basically the same, perhaps with a clearer syntax:
```groovy
params {
    // no glob patterns, only simple names
    withName: SRATOOLS_FASTERQDUMP {
        args_fasterqdump = '--split-files --include-technical'
    }
}
```

So it would be basically the same approach as the ext config, but since the process declares these settings, Nextflow can validate them in the config, provide auto-completion, etc.

My only concern is that these "process params" could be abused. At the same time, these params can basically only be strings, so that limits their usability to things like args, prefix, etc, so maybe it doesn't matter. Need to think on it more

mahesh-panchal replied:
I agree with the potential for abuse, but the ability to use closures has been a very strong attraction here in the usage of ext.args. It's essentially a necessity for `ext.prefix` and highly desired to be able to select parameters based on pipeline `input:`s too. 

Something I'm wondering is if a profiles approach might also solve this issue ( although it would have to be something that supports nesting to some degree ) so it could be set using a json schema or something but this might be too complex for the average user.

mahesh-panchal replied:
Also using only simple names is not going to help, for example nf-core rnaseq has 5 patterns for `SAMTOOLS_INDEX`

mahesh-panchal replied:
Also I fully support having this in params though because anything that could potentially change the results should be in the `-params-file` and not half in params and half in process such as a `nextflow.config` in the launch directory which could be forgotten if someone comes back later to rerun something.

mahesh-panchal replied:
I've also come to the opinion that it's likely only the `ext.args` equivalent that should be in the params. There are parts that should remain under pipeline developer control, so having these params as strings is perhaps fine, as the dynamic part should then perhaps be process inputs instead, on which these user defined extra opts get appended. The `ext.prefix` is also something that should likely be a process `input:` too rather than something a user can control. 

```nextflow
process FOO {
    input:
    path bar
    val args, params: true // Map where keys in the map are appendable from params, otherwise throws an error?
    val prefix // Map [ bwa_prefix, samtools_prefix ]

    script:
    """
    ...
    """
}
```
What worries me about this idea is that a user will say they want to remove developer set args ( where the idea is the developer only sets what's necessary for pipeline stability ).

bentsherman replied:
Following the latest refactor, I think it will be possible to declare ext args as regular inputs:
```groovy
  input:
  // ...
  args_fasterqdump: String
  args_pigz: String
  prefix: String
```

But now, these inputs could be supplied either in the pipeline code:
```groovy
ch_input.map(FOO, args_pigz: '...')
```

Or the config:
```groovy
process {
    withName: FOO {
        ext.args_fasterqdump = '--split-files --include-technical'
    }
}
```

This becomes possible because extra process inputs are now specified by name in the pipeline code. The only question would be how to handle overrides -- it wouldn't make sense for the user to be able to override "main" inputs like sample id from the config, so might still need a way to distinguish "main" inputs from "extra" inputs.

mahesh-panchal replied:
If one could override an entire map from the config, that would mean people could also use portions of a workflow too. So I can see a use for overriding from the config. 

I like the idea though, but there definitely needs to be something to handle the overrides.

bentsherman commented on workflows/sra/main.nf:
```diff
@@ -29,164 +27,119 @@ include { FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS } from '../../subworkflow
-    sample_mappings = ch_sample_mappings_yml
-    sra_metadata    = ch_sra_metadata
-    versions        = ch_versions.unique()
+    ftp_samples
+        .mix(sratools_samples)
+        .mix(aspera_samples)
```
Instead of doing a separate filter/map for each branch, you could handle all three branches in a single map operator:
```groovy
ncbi_settings = params.download_method == DownloadMethod.SRATOOLS
    ? CUSTOM_SRATOOLSNCBISETTINGS( sra_metadata.collect() )
    : null

samples = sra_metadata.map { meta ->
    def method = getDownloadMethod(meta, params.download_method)
    match( method ) { // or switch, or if-else, etc
        DownloadMethod.FTP -> {
            def out = SRA_FASTQ_FTP ( meta )
            new Sample(meta.id, out.fastq_1, out.fastq_2, out.md5_1, out.md5_2)
        },
        DownloadMethod.SRATOOLS -> {
            def sra = SRATOOLS_PREFETCH ( meta, ncbi_settings, dbgap_key )
            def fastq = SRATOOLS_FASTERQDUMP ( meta, sra, ncbi_settings, dbgap_key )
            def fastq_1 = fastq[0]
            def fastq_2 = !meta.single_end ? fastq[1] : null
            new Sample(meta.id, fastq_1, fastq_2, null, null)
        },
        DownloadMethod.ASPERA -> {
            def out = ASPERA_CLI ( meta, 'era-fasp' )
            new Sample(meta.id, out.fastq_1, out.fastq_2, out.md5_1, out.md5_2)
        }
    }
}
```

(the `match` is like a switch-case or if-else chain, just an idea I had for how we could do it in Nextflow)

I think this is a more concise description of the workflow, because you basically just write "regular code" and less operator logic. Now I enjoy playing with operators as much as anyone, but beyond a certain point they become a distraction in my opinion. Others may disagree.

On the other hand, this approach isn't very amenable to subworkflows. You can't call a workflow inside the map closure, and you can't factor out any of this code into a separate function. I might be able to remove this limitation in the future. But for now I think this approach would lead to fewer subworkflows. Maybe that's fine if the refactor simplifies the overall workflow logic by a lot. You certainly wouldn't have to argue so much about naming workflows then 😉 

At the end of the day, either way works just fine. You can split up the logic and use more operators, which allows you to factor out subworkflows like for the SRATOOLS route, or you can do it all in one giant map operator, or something in between.

Either way is easy to do for fetchngs because you just *mix* the various routes at the end. But when you need to *join* several related channels from different processes -- think like raw fastqc + trimmed fastqc + bam for each sample in rnaseq -- I think that's where the giant map operator is much easier. A good rule of thumb might be to use separate operators for mixing, but use a single map operator for joining.

mahesh-panchal replied:
I think it would be nice if subworkflows behaved in a similar fashion to processes. At the moment, they're just zip-ties for channels, and I think people would find working with subworkflows nicer if they behaved more like functions.

bentsherman replied:
On the topic of subworkflows, do you think people should be able to do things like channels of channels? Because that's what making subworkflows more like functions would mean. You would be able to have nested levels of concurrency, and you wouldn't be able to know the full process DAG up front.

bentsherman replied:
I guess not knowing the full process DAG isn't inherently bad, but it would be a fundamental change in Nextflow's execution model.

But my bigger concern is that it might make Nextflow much more complicated -- basically on the same level as any general-purpose language -- which might be prohibitive for many potential developers.

I mean, if we're talking about extra process inputs being too much of a hassle, imagine what allowing channels of channels would do 😅

mahesh-panchal replied:
Hmm, I hadn't thought of channels of channels. It's not something I think should be permitted. My thought was really more like passing a list input which would get turned into a channel, explicitly in the subworkflow. Like a process, a subworkflow wouldn't take a channel as input, but you would use map to call it on groups of things most likely.

mahesh-panchal replied:
I was thinking about the subworkflows only allowing non-channel inputs suggestion again (in this DSL3 proposal context) and realized might this make recursion and feedback loops easier to implement? If a subworkflow has to treat inputs and outputs a bit like processes (i.e. all inputs and outputs are a non-channel type, and inputs are explicitly converted into channels in the `main:` block with channel factories), might it make certain things easier to compose, or does it just become memory hungry with all the extra objects that get created (I'm thinking if someone spawns 10,000 subworkflows for some purpose rather than the handful of channels if it remains like a zip-tie).

bentsherman replied:
After the latest refactor, my original comment here (the giant match-case block) is no longer relevant. The existing paradigm of operators and workflows is largely retained, only simplified.

I continue to see subworkflows as essentially "compiler macros" (what you call "zip-ties for channels"). But I am keeping in the back of my head the idea of a function that *doesn't* allow channel inputs but *does* allow some kind of nested parallelization via processes.

mahesh-panchal replied:
I have to admit, my naive thinking of "subworkflows as functions" I don't think scales very well, so the current model is still necessary. 

It's more there needs to be some kind of translation from it, because people do find thinking in that way a bit easier, and translates to recursion easier too. There seems to be interest in as seen here https://nextflow.slack.com/archives/C02T98A23U7/p1745459250690079

mahesh-panchal replied:
Would it be conceivable to have another type of block different from `workflow`? A `composite` block for lack of words atm. The idea being that those processes are not linked with channels, but just pass values from one process to the next (There can still be channels under the hood). This might also make for logical grouping for those that want a group submission feature. This would be something that could be recursed on too, rather than a `workflow`. I think this might be a very popular feature.

bentsherman replied:
I think this could work. For example, we could have a `workflow` that uses `input/output` instead of `take/emit`:

```groovy
workflow SRATOOLS {
  input:
  meta: Map
  ncbi_settings: Path
  dbgap_key: String
  
  main:
  sra = SRATOOLS_PREFETCH ( meta, ncbi_settings, dbgap_key )
  fastq = SRATOOLS_FASTERQDUMP ( meta, sra, ncbi_settings, dbgap_key )

  output:
  sample: Sample = meta + fastq
}

record Sample {
  id: String
  // ...
}
``` 

Or, alternatively, a `process` that uses `main` instead of `script`, and can compose other processes:
```groovy
process SRATOOLS {
  input:
  // (same as above)
  
  main:
  sra = SRATOOLS_PREFETCH ( meta, ncbi_settings, dbgap_key )
  fastq = SRATOOLS_FASTERQDUMP ( meta, sra, ncbi_settings, dbgap_key )

  output:
  // (same as above)
}
```

bentsherman commented:
I found a way to achieve static types with significantly fewer changes. and still managed to simplify the dataflow logic overall.

The key breakthrough is to treat the record type as a contract over a map (like the meta-map), rather than a fixed data type that can only contain the record members at all times. This gives you a great deal of flexibility.

We still use a `map` operator to call a process. But instead of calling the process *in* the operator closure, we call it *as* the operator closure. For each map in the channel, the map keys are matched to the process inputs by name.
- The map can have additional keys, the process just won't use them
- If the map is missing some keys, they can be provided as named args (no need to use `join` or `combine`)

Record types are not used much explicitly. They are mainly used in input/output declarations, but most of the intermediate code is just channels of maps.

This turns out to be really elegant, and it will be much easier for me to implement 😅

mahesh-panchal commented:
Re: the reduced operator library. 

This is an interesting use case for `collectFile`:
> With that many files you need manage throughput/load both on the NF side, the lustre side, and maybe the SLURM side if that is your executor. 50M files (if I read that correctly - 42 seconds seems quite fast to find that many) is a lot. I dealt with millions of files on lustre about 6-7 years ago.  It is doable, but you’re putting a big load on the system, you’ll be affected by the rest of the load and the rest of load will be affected by you. I also used this trick to avoid staging large amounts of files by NF in the task directory:
>
> ```
> map { it[0] + "\t" + it[1].toString() + "\n" }.collectFile(name: 'stats.meta')
> ```
> In that case the input was a channel of (tuple) `[val(sampleID), path(somefile)]` . Normally NF would stage (each) path as a symlink in the task directory; with the above you just get a meta/manifest file that lists all the locations.

Context: https://nextflow.slack.com/archives/C02T98A23U7/p1752236423963939?thread_ts=1752224037.656359&cid=C02T98A23U7

Does the reduced operator library mean the developer would have to use a process first each time to create a manifest? We can't just use map to create a file since it should go into a working directory rather than launch directory.

bentsherman commented:
My idea for `collectFile` is to provide a generic function for writing text/csv files that can be composed with `groupTuple` and `List::sort` as needed to achieve the same functionality, like in [this example](https://github.com/bentsherman/nf-boost/blob/main/examples/merge-text.nf).

In that example I do use a process, but you could just do it in a `map` operator if you want to avoid making a bunch of task directories. You just lose the automatic retry/caching, but maybe you don't need that. So that use case should be totally feasible with the reduced operators.

bentsherman committed:
Upgrade to Nextflow 26.04 [wip]
